src/main/scala/se/nimsa/sbx/filtering/FilteringDAO.scala

```diff
@@ -0,0 +1,196 @@
+  }
+
+  object TagFilterTable {
+    val name = "TagFilters"
+  }
+
+  val tagPathQuery = TableQuery[TagPathTable]
+
+  class TagPathTable(tag: Tag) extends Table[TagFilterTagPath](tag, TagPathTable.name) {
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def tagFilterId = column[Long]("tagfilterid")
+    def tagPath = column[String]("tagpath")
+    def fkTagFilter = foreignKey("fk_tag_filter", tagFilterId, tagFilterQuery)(_.id, onDelete = ForeignKeyAction.Cascade)
+    def * = (id, tagFilterId, tagPath) <> (toTagFilterTagPath.tupled, (a: TagFilterTagPath) => Option(a.id, a.tagFilterId, a.tagPathTag.toString()))
+  }
+
+  object TagPathTable {
+    val name = "TagPaths"
+  }
+
+  val sourceFilterQuery = TableQuery[SourceFilterTable]
+
+  class SourceFilterTable(tag: Tag) extends Table[SourceTagFilter](tag, SourceFilterTable.name) {
+    def id = column[Long]("id", O.PrimaryKey, O.AutoInc)
+    def sourceType = column[String]("sourcetype", O.Length(64))
+    def sourceId = column[Long]("sourceid")
+    def tagFilterId = column[Long]("tagfilterid")
+    def fkTagFilter2 = foreignKey("fk_tag_filter2", tagFilterId, tagFilterQuery)(_.id, onDelete = ForeignKeyAction.Cascade)
+    def * = (id, sourceType, sourceId, tagFilterId) <> (toSourceTagFilter.tupled, (a: SourceTagFilter) => Option(a.id, a.sourceType.toString, a.sourceId, a.tagFilterId))
+  }
+
+  object SourceFilterTable {
+    val name = "SourceTagFilters"
+  }
+
+  def create() = createTables(dbConf, (TagFilterTable.name, tagFilterQuery), (TagPathTable.name, tagPathQuery),
+    (SourceFilterTable.name, sourceFilterQuery))
+
+  def drop() = db.run {
+    (tagFilterQuery.schema ++ tagPathQuery.schema ++ sourceFilterQuery.schema).drop
+  }
+
+  def clear() = db.run {
+    DBIO.seq(tagFilterQuery.delete, tagPathQuery.delete, sourceFilterQuery.delete)
+  }
+
+  def dump() = db.run {
+    val fullOuterJoin = for {
+      (c, s) <- tagFilterQuery joinFull tagPathQuery on (_.id === _.tagFilterId)
+    } yield (c.map(_.name), s.map(_.tagPath))
+    fullOuterJoin.result
+  }
+
+  def createOrUpdateTagFilter(tagFilter: TagFilterSpec): Future[TagFilterSpec] = {
+    val tagFilterRow = TagFilter(tagFilter.id, tagFilter.name, tagFilter.tagFilterType)
+    val insertAction = for {
+      tfr <- insertTagFilterAction(tagFilterRow)
+      bbr <- replaceTagFilterTagPathAction(tagFilter.tags.map(tftp => TagFilterTagPath(-1, tfr.id, tftp)))
+    } yield (tfr)
+    val res = db.run {
+      getTagFilterByNameAction(tagFilter.name).flatMap {
+        _.map {t =>
+          replaceTagFilterTagPathAction(tagFilter.tags.map(tftp => TagFilterTagPath(-1, t.id, tftp))) andThen
+          updateTagFilterAction(tagFilterRow).map(_ => tagFilterRow.copy(id = t.id))
+        }.getOrElse {
+          insertAction
+        }
+      }
+    }
+    res.map(r => {
+      val res = tagFilter.copy(id = r.id)
+      res
+    })
```

Comment:
I see several places in this file where two or more futures  are combined. It would be great if this can be expressed in DBIO actions instead and run in a single future. That way we can run things requiring multiple database calls in a transaction and minimize the risk of having half-baked results in the database. DBIO actions can be composed just like futures.

Reply:
I don't see where, can you give me a pointer.

Reply:
It is combined in the `res.map(...`. You can typically do this mapping on the DB actions instead of on the futures. Then, the chain of DB actions can be run in a transaction by `db.run(...).transactionally`.